src/hll.c

```diff
@@ -92,6 +112,318 @@ enum {
+void _PG_init(void)
+{
+	if (!process_shared_preload_libraries_in_progress)
+	{
+		ereport(ERROR, (errmsg("HLL can only be loaded via shared_preload_libraries"),
+						errhint("Add hll to shared_preload_libraries configuration "
+								"variable in postgresql.conf")));
+	}
+
+	/*
+	 * Register HLL configuration variables.
+	 */
+	RegisterHLLConfigVariables();
+
+	previous_upper_path_hook = create_upper_paths_hook;
+	create_upper_paths_hook = hll_aggregation_restriction_hook;
+}
+
+/*
+ * hll_aggregation_restriction_hook is assigned to create_upper_paths_hook to
+ * check whether there exist a path with hash aggregate. If that aggregate is
+ * introduced by hll, it's cost is maximized to force planner to not to select
+ * hash aggregate.
+ *
+ * Since the signature of the hook changes after PG 11, we define the signature
+ * and the previous hook call part of this function depending on the PG version.
+ */
+static void
+#if (PG_VERSION_NUM >= 110000)
+
+hll_aggregation_restriction_hook(PlannerInfo *root, UpperRelationKind stage,
+								   RelOptInfo *input_rel, RelOptInfo *output_rel,
+								   void *extra)
+#else
+hll_aggregation_restriction_hook(PlannerInfo *root, UpperRelationKind stage,
+								   RelOptInfo *input_rel, RelOptInfo *output_rel)
+#endif
+{
+	Oid hllId = InvalidOid;
+
+	// If previous hook exist, call it first to get most update path
+	if (previous_upper_path_hook != NULL)
+	{
+		#if (PG_VERSION_NUM >= 110000)
+			previous_upper_path_hook(root, stage, input_rel, output_rel, extra);
+		#else
+			previous_upper_path_hook(root, stage, input_rel, output_rel);
+		#endif
+	}
+
+	/* If HLL extension is not loaded, do nothing */
+	hllId = get_extension_oid(HLL_EXTENSION_NAME, true);
+	if (!OidIsValid(hllId))
+	{
+		return;
+	}
+
+	/*
+	 *  If the client force the group agg, maximize the cost of the path with
+	 *  the hash agg to force planner to choose group agg instead.
+	 */
+	if (ForceGroupagg)
+	{
+		if (stage == UPPERREL_GROUP_AGG || stage == UPPERREL_FINAL)
+		{
+			ListCell *pathCell = list_head(output_rel->pathlist);
+			foreach(pathCell, output_rel->pathlist)
+			{
+				Path *path = (Path *) lfirst(pathCell);
+				if(path->pathtype == T_Agg && ((AggPath *)path)->aggstrategy == AGG_HASHED)
+				{
+					MaximizeCostOfHashAggregateWithHLL(path);
+				}
+			}
+		}
+	}
+}
+
+/*
+ * MaximizeCostOfHashAggregateWithHLL maximizes the cost of the path if it tries
+ * to run hll aggregate function with hash aggregate.
+ */
+static void
+MaximizeCostOfHashAggregateWithHLL(Path *path)
+{
+	List *varList = pull_var_clause((Node*) path->pathtarget->exprs,
+									PVC_INCLUDE_AGGREGATES);
+	ListCell *varCell = NULL;
+
+	foreach(varCell, varList)
+	{
+		Var *var = (Var *) lfirst(varCell);
+
+		if (nodeTag(var) == T_Aggref)
+		{
+			Aggref *aggref = (Aggref *) var;
+			int argCount = list_length(aggref->args);
+			Oid addFunctionId = LookupAggregatorOidByName(HLL_ADD_AGG_ID, argCount);
```

Comment:
I'm starting to think, that it would be better to have two static variables (like add_aggregate_oid[] and union_aggregate_id) for function oids which are populated at pg_init. This would save us from all caching logic.